File: cs/src/DataCentric/Platform/Storage/DataSourceData.cs
/*
Copyright (C) 2013-present The DataCentric Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

using System;
using System.Collections.Generic;
using MongoDB.Bson.Serialization.Attributes;
using NodaTime;

namespace DataCentric
{
    /// <summary>
    /// Data source is a logical concept similar to database
    /// that can be implemented for a document DB, relational DB,
    /// key-value store, or filesystem.
    ///
    /// Data source API provides the ability to:
    ///
    /// (a) store and query datasets;
    /// (b) store records in a specific dataset; and
    /// (c) query record across a group of datasets.
    ///
    /// This record is stored in root dataset.
    /// </summary>
    public abstract class DataSourceData : RootRecord<DataSourceKey, DataSourceData>, IDataSource
    {
        private Dictionary<string, RecordId> dataSetDict_ { get; } = new Dictionary<string, RecordId>();
        private Dictionary<RecordId, HashSet<RecordId>> importDict_ { get; } = new Dictionary<RecordId, HashSet<RecordId>>();

        //--- ELEMENTS

        /// <summary>Unique data source name.</summary>
        [BsonRequired]
        public string DataSourceName { get; set; }

        /// <summary>
        /// This class enforces strict naming conventions
        /// for database naming. While format of the resulting database
        /// name is specific to data store type, it always consists
        /// of three tokens: InstanceType, InstanceName, and EnvName.
        /// The meaning of InstanceName and EnvName tokens depends on
        /// the value of InstanceType enumeration.
        /// </summary>
        public DbNameKey DbName { get; set; }

        /// <summary>
        /// Use this flag to mark data source as readonly.
        ///
        /// This value will be set to true if CutoffTime is
        /// true, because a data source that views the data
        /// for a cutoff time cannot modify its view.
        /// </summary>
        public bool? ReadOnly { get; set; }

        /// <summary>
        /// Records where _id is greater than SavedById will be
        /// ignored by the data source.
        /// </summary>
        public RecordId? SavedBy { get; set; }

        //--- METHODS

        /// <summary>
        /// Releases resources and calls base.Dispose().
        ///
        /// This method will not be called by the garbage collector.
        /// It will only be executed if:
        ///
        /// * This class implements IDisposable; and
        /// * The class instance is created through the using clause
        ///
        /// IMPORTANT - Every override of this method must call base.Dispose()
        /// after executing its own code.
        /// </summary>
        public virtual void Dispose()
        {
            // Uncomment except in root class of the hierarchy
            // base.Dispose();
        }

        /// <summary>Flush data to permanent storage.</summary>
        public abstract void Flush();

        /// <summary>
        /// The returned RecordIds have the following order guarantees:
        ///
        /// * For this data source instance, to arbitrary resolution; and
        /// * Across all processes and machines, to one second resolution
        ///
        /// One second resolution means that two RecordIds created within
        /// the same second by different instances of the data source
        /// class may not be ordered chronologically unless they are at
        /// least one second apart.
        /// </summary>
        public abstract RecordId CreateOrderedRecordId();

        /// <summary>
        /// Load record by its RecordId.
        ///
        /// Return null if there is no record for the specified RecordId;
        /// however an exception will be thrown if the record exists but
        /// is not derived from TRecord.
        /// </summary>
        public abstract TRecord LoadOrNull<TRecord>(RecordId id)
            where TRecord : Record;

        /// <summary>
        /// Load record by string key from the specified dataset or
        /// its list of imports. The lookup occurs first in descending
        /// order of dataset RecordIds, and then in the descending
        /// order of record RecordIds within the first dataset that
        /// has at least one record. Both dataset and record RecordIds
        /// are ordered chronologically to one second resolution,
        /// and are unique within the database server or cluster.
        ///
        /// The root dataset has empty RecordId value that is less
        /// than any other RecordId value. Accordingly, the root
        /// dataset is the last one in the lookup order of datasets.
        ///
        /// The first record in this lookup order is returned, or null
        /// if no records are found or if DeletedRecord is the first
        /// record.
        ///
        /// Return null if there is no record for the specified RecordId;
        /// however an exception will be thrown if the record exists but
        /// is not derived from TRecord.
        /// </summary>
        public abstract TRecord LoadOrNull<TKey, TRecord>(TypedKey<TKey, TRecord> key, RecordId loadFrom)
            where TKey : TypedKey<TKey, TRecord>, new()
            where TRecord : TypedRecord<TKey, TRecord>;

        /// <summary>
        /// Get query for the specified type.
        ///
        /// After applying query parameters, the lookup occurs first in
        /// descending order of dataset RecordIds, and then in the descending
        /// order of record RecordIds within the first dataset that
        /// has at least one record. Both dataset and record RecordIds
        /// are ordered chronologically to one second resolution,
        /// and are unique within the database server or cluster.
        ///
        /// The root dataset has empty RecordId value that is less
        /// than any other RecordId value. Accordingly, the root
        /// dataset is the last one in the lookup order of datasets.
        ///
        /// Generic parameter TRecord is not necessarily the root data type;
        /// it may also be a type derived from the root data type.
        /// </summary>
        public abstract IQuery<TRecord> GetQuery<TRecord>(RecordId loadFrom)
            where TRecord : Record;

        /// <summary>
        /// Save record to the specified dataset. After the method exits,
        /// record.DataSet will be set to the value of the dataSet parameter.
        ///
        /// All Save methods ignore the value of record.DataSet before the
        /// Save method is called. When dataset is not specified explicitly,
        /// the value of dataset from the context, not from the record, is used.
        /// The reason for this behavior is that the record may be stored from
        /// a different dataset than the one where it is used.
        ///
        /// This method guarantees that RecordIds will be in strictly increasing
        /// order for this instance of the data source class always, and across
        /// all processes and machine if they are not created within the same
        /// second.
        /// </summary>
        public abstract void Save<TRecord>(TRecord record, RecordId saveTo)
            where TRecord : Record;

        /// <summary>
        /// Write a DeletedRecord in deleteIn dataset for the specified key
        /// instead of actually deleting the record. This ensures that
        /// a record in another dataset does not become visible during
        /// lookup in a sequence of datasets.
        ///
        /// To avoid an additional roundtrip to the data store, the delete
        /// marker is written even when the record does not exist.
        /// </summary>
        public abstract void Delete<TKey, TRecord>(TypedKey<TKey, TRecord> key, RecordId deleteIn)
            where TKey : TypedKey<TKey, TRecord>, new()
            where TRecord : TypedRecord<TKey, TRecord>;

        /// <summary>
        /// Permanently deletes (drops) the database with all records
        /// in it without the possibility to recover them later.
        ///
        /// This method should only be used to free storage. For
        /// all other purposes, methods that preserve history should
        /// be used.
        ///
        /// ATTENTION - THIS METHOD WILL DELETE ALL DATA WITHOUT
        /// THE POSSIBILITY OF RECOVERY. USE WITH CAUTION.
        /// </summary>
        public abstract void DeleteDb();

        //--- METHODS

        /// <summary>
        /// Error message if the data source is readonly.
        ///
        /// This method also provides an alert if CutoffTime
        /// is set but ReadOnly flag is not because  is not
        /// possible to write to a view of the data as of a
        /// past point in time.
        /// </summary>
        public void CheckNotReadOnly()
        {
            if (ReadOnly != null && ReadOnly.Value)
                throw new Exception(
                    $"Attempting write operation for readonly data source {DataSourceName}. " +
                    $"A data source is readonly if either (a) its ReadOnly flag is set, or (b) " +
                    $"one of SavedByTime or SavedById is set.");
            else if (SavedBy != null)
                throw new Exception(
                    $"Data source {DataSourceName} has CutoffTime set, but is not ReadOnly. " +
                    $"It is not possible to write to a view of the data as of a past point in time.");
        }

        /// <summary>
        /// Get RecordId of the dataset with the specified name.
        ///
        /// All of the previously requested dataSetIds are cached by
        /// the data source. To load the latest version of the dataset
        /// written by a separate process, clear the cache first by
        /// calling DataSource.ClearDataSetCache() method.
        ///
        /// Returns null if not found.
        /// </summary>
        public RecordId? GetDataSetOrNull(string dataSetName, RecordId loadFrom)
        {
            if (dataSetDict_.TryGetValue(dataSetName, out RecordId result))
            {
                // Check if already cached, return if found
                return result;
            }
            else
            {
                // Otherwise load from storage (this also updates the dictionaries)
                DataSetKey dataSetKey = new DataSetKey() { DataSetName = dataSetName };
                DataSetData dataSetData = this.LoadOrNull(dataSetKey, loadFrom);

                // If not found, return RecordId.Empty
                if (dataSetData == null) return null;

                // If found, cache result in RecordId dictionary
                dataSetDict_[dataSetName] = dataSetData.Id;

                // Build and cache dataset lookup list if not found
                if (!importDict_.TryGetValue(dataSetData.Id, out HashSet<RecordId> importSet))
                {
                    importSet = BuildDataSetLookupList(dataSetData);
                    importDict_.Add(dataSetData.Id, importSet);
                }

                return dataSetData.Id;
            }
        }

        /// <summary>
        /// Save new version of the dataset.
        ///
        /// This method sets Id element of the argument to be the
        /// new RecordId assigned to the record when it is saved.
        /// The timestamp of the new RecordId is the current time.
        ///
        /// This method updates in-memory cache to the saved dataset.
        /// </summary>
        public void SaveDataSet(DataSetData dataSetData, RecordId saveTo)
        {
            // Save dataset to storage. This updates its Id
            // to the new RecordId created during save
            Save<DataSetData>(dataSetData, saveTo);

            // Update dataset dictionary with the new Id
            dataSetDict_[dataSetData.Key] = dataSetData.Id;

            // Update lookup list dictionary
            var lookupList = BuildDataSetLookupList(dataSetData);
            importDict_.Add(dataSetData.Id, lookupList);
        }

        /// <summary>
        /// Returns enumeration of import datasets for specified dataset data,
        /// including imports of imports to unlimited depth with cyclic
        /// references and duplicates removed.
        ///
        /// The list will not include datasets that are after the value of
        /// CutoffTime if specified, or their imports (including
        /// even those imports that are earlier than the constraint).
        /// </summary>
        public IEnumerable<RecordId> GetDataSetLookupList(RecordId loadFrom)
        {
            // Root dataset has no imports (there is not even a record
            // where these imports can be specified).
            //
            // Return list containing only the root dataset (RecordId.Empty) and exit
            if (loadFrom == RecordId.Empty)
            {
                return new RecordId[] {RecordId.Empty};
            }

            if (importDict_.TryGetValue(loadFrom, out HashSet<RecordId> result))
            {
                // Check if the lookup list is already cached, return if yes
                return result;
            }
            else
            {
                // Otherwise load from storage (returns null if not found)
                DataSetData dataSetData = LoadOrNull<DataSetData>(loadFrom);

                if (dataSetData == null) throw new Exception($"Dataset with RecordId={loadFrom} is not found.");
                if (dataSetData.DataSet != RecordId.Empty) throw new Exception($"Dataset with RecordId={loadFrom} is not stored in root dataset.");

                // Build the lookup list
                result = BuildDataSetLookupList(dataSetData);

                // Add to dictionary and return
                importDict_.Add(loadFrom, result);
                return result;
            }
        }

        //--- PRIVATE

        /// <summary>
        /// Builds hashset of import datasets for specified dataset data,
        /// including imports of imports to unlimited depth with cyclic
        /// references and duplicates removed. This method uses cached lookup
        /// list for the import datasets but not for the argument dataset.
        ///
        /// The list will not include datasets that are after the value of
        /// CutoffTime if specified, or their imports (including
        /// even those imports that are earlier than the constraint).
        ///
        /// This overload of the method will return the result hashset.
        ///
        /// This private helper method should not be used directly.
        /// It provides functionality for the public API of this class.
        /// </summary>
        private HashSet<RecordId> BuildDataSetLookupList(DataSetData dataSetData)
        {
            // Delegate to the second overload
            var result = new HashSet<RecordId>();
            BuildDataSetLookupList(dataSetData, result);
            return result;
        }

        /// <summary>
        /// Builds hashset of import datasets for specified dataset data,
        /// including imports of imports to unlimited depth with cyclic
        /// references and duplicates removed. This method uses cached lookup
        /// list for the import datasets but not for the argument dataset.
        ///
        /// The list will not include datasets that are after the value of
        /// CutoffTime if specified, or their imports (including
        /// even those imports that are earlier than the constraint).
        ///
        /// This overload of the method will return the result hashset.
        ///
        /// This private helper method should not be used directly.
        /// It provides functionality for the public API of this class.
        /// </summary>
        private void BuildDataSetLookupList(DataSetData dataSetData, HashSet<RecordId> result)
        {
            // Return if the dataset is null or has no imports
            if (dataSetData == null) return;

            // Error message if dataset has no Id or Key set
            dataSetData.Id.CheckHasValue();
            dataSetData.Key.CheckHasValue();

            if (SavedBy != null && dataSetData.Id > SavedBy.Value)
            {
                // Do not add if revision time constraint is set and is before this dataset.
                // In this case the import datasets should not be added either, even if they
                // do not fail the revision time constraint
                return;
            }

            // Add self to the result
            result.Add(dataSetData.Id);

            // Add imports to the result
            if (dataSetData.Imports != null)
            {
                foreach (var dataSetId in dataSetData.Imports)
                {
                    // Dataset cannot include itself as its import
                    if (dataSetData.Id == dataSetId)
                        throw new Exception(
                            $"Dataset {dataSetData.Key} with RecordId={dataSetData.Id} includes itself in the list of its imports.");

                    // The Add method returns true if the argument is not yet present in the hashset
                    if (result.Add(dataSetId))
                    {
                        // Add recursively if not already present in the hashset
                        var cachedImportList = GetDataSetLookupList(dataSetId);
                        foreach (var importId in cachedImportList)
                        {
                            result.Add(importId);
                        }
                    }
                }
            }
        }
    }
}

File: py/src/datacentric/platform/storage/data_source_data.py
from abc import ABC, abstractmethod
from bson.objectid import ObjectId
from typing import List, Set, Dict, Iterable, Optional, TypeVar

from datacentric.platform.storage.db_name import DbNameKey
from datacentric.types.record import RootRecord, TypedKey, Record
from datacentric.platform.data_set import DataSetData, DataSetKey

TRecord = TypeVar('TRecord', bound=Record)


class DataSourceKey(TypedKey['DataSourceData']):
    """Key class for DataSourceData.
    Record associated with this key is stored in root dataset.
    """
    __slots__ = ['data_source_name']

    data_source_name: str
    cache: str = 'Cache'
    master: str = 'Master'

    def __init__(self, value: str = None):
        super().__init__()
        self.data_source_name = None
        if value is not None:
            self.data_source_name = value


class DataSourceData(RootRecord[DataSourceKey], ABC):
    """Data source is a logical concept similar to database
    that can be implemented for a document DB, relational DB,
    key-value store, or filesystem.

    Data source API provides the ability to:

    (a) store and query datasets;
    (b) store records in a specific dataset; and
    (c) query record across a group of datasets.

    This record is stored in root dataset.
    """

    __slots__ = ('data_source_name', 'db_name', 'non_temporal', 'readonly', '_data_set_dict', '_import_dict')

    _empty_id = ObjectId('000000000000000000000000')
    common_id: str = 'Common'

    _data_set_dict: Dict[str, ObjectId]
    _import_dict: Dict[ObjectId, Set[ObjectId]]

    data_source_name: str
    """Unique data source name."""

    db_name: DbNameKey
    """Database name."""

    non_temporal: bool
    """For the data stored in data sources where non_temporal == false,
    the data source keeps permanent history of changes to each
    record (except where dataset or record are marked as NonTemporal),
    and provides the ability to access the record as of the specified
    TemporalId, where TemporalId serves as a timeline (records created
    later have greater TemporalId than records created earlier).
    
    For the data stored in data source where NonTemporal == true,
    the data source keeps only the latest version of the record. All
    datasets created by a NonTemporal data source must also be non-temporal.
    
    In a non-temporal data source, this flag is ignored as all
    datasets in such data source are non-temporal.
    """

    readonly: bool
    """Use this flag to mark data source as readonly."""

    def __init__(self):
        super().__init__()
        self._data_set_dict = dict()
        self._import_dict = dict()

        self.data_source_name = None
        self.db_name = None
        self.non_temporal = None
        self.readonly = None

    @abstractmethod
    def create_ordered_object_id(self) -> ObjectId:
        """The returned ObjectIds have the following order guarantees:

        * For this data source instance, to arbitrary resolution; and
        * Across all processes and machines, to one second resolution

        One second resolution means that two ObjectIds created within
        the same second by different instances of the data source
        class may not be ordered chronologically unless they are at
        least one second apart."""
        pass

    @abstractmethod
    def is_readonly(self) -> bool:
        pass

    def check_not_readonly(self):
        if self.is_readonly():
            raise Exception(f'Attempting write operation for readonly data source {self.data_source_name}. '
                            f'A data source is readonly if either (a) its ReadOnly flag is set, or (b) '
                            f'one of SavedByTime or SavedById is set.')

    @abstractmethod
    def load_or_null(self, id_: ObjectId, type_: type) -> Optional[TRecord]:
        """Load record by its ObjectId.

        Return None if there is no record for the specified ObjectId;
        however an exception will be thrown if the record exists but
        is not derived from type_.
        """
        pass

    @abstractmethod
    def load_or_null_by_key(self, key_: TypedKey, load_from: ObjectId) -> Optional[TRecord]:
        """Load record by string key from the specified dataset or
        its list of imports. The lookup occurs first in descending
        order of dataset ObjectIds, and then in the descending
        order of record ObjectIds within the first dataset that
        has at least one record. Both dataset and record ObjectIds
        are ordered chronologically to one second resolution,
        and are unique within the database server or cluster.

        The root dataset has empty ObjectId value that is less
        than any other ObjectId value. Accordingly, the root
        dataset is the last one in the lookup order of datasets.

        The first record in this lookup order is returned, or null
        if no records are found or if DeletedRecord is the first
        record.

        Return None if there is no record for the specified ObjectId;
        however an exception will be thrown if the record exists but
        is not derived from TRecord.
        """
        pass

    @abstractmethod
    def get_query(self, load_from: ObjectId, type_: type):
        pass

    @abstractmethod
    def save(self, record: TRecord, save_to: ObjectId) -> None:
        pass

    @abstractmethod
    def delete(self, key: TypedKey[TRecord], delete_in: ObjectId) -> None:
        pass

    @abstractmethod
    def delete_db(self) -> None:
        pass

    def get_data_set_or_none(self, data_set_id: str, load_from: ObjectId) -> Optional[ObjectId]:
        """Get ObjectId of the dataset with the specified name.

        All of the previously requested data_set_ids are cached by
        the data source.

        Returns null if not found.
        """
        if data_set_id in self._data_set_dict:
            return self._data_set_dict[data_set_id]
        else:
            data_set_key = DataSetKey(data_set_id)

            data_set_data: DataSetData = self.load_or_null_by_key(data_set_key, load_from)

            if data_set_data is None:
                return None

            # If found cache in dictionary
            self._data_set_dict[data_set_id] = data_set_data.id_
            if data_set_data.id_ not in self._import_dict:
                import_set = self.__build_data_set_lookup_list(data_set_data)
                self._import_dict[data_set_data.id_] = import_set
            return data_set_data.id_

    def save_data_set(self, data_set_data: DataSetData, save_to: ObjectId) -> None:
        """Save new version of the dataset and update in-memory cache to the saved dataset."""
        self.save(data_set_data, save_to)
        self._data_set_dict[data_set_data.key] = data_set_data.id_
        lookup_list = self.__build_data_set_lookup_list(data_set_data)
        self._import_dict[data_set_data.id_] = lookup_list

    def get_data_set_lookup_list(self, load_from: ObjectId) -> Iterable[ObjectId]:
        """Returns enumeration of import datasets for specified dataset data,
        including imports of imports to unlimited depth with cyclic
        references and duplicates removed.
        """
        if load_from == DataSourceData._empty_id:
            return [DataSourceData._empty_id]
        if load_from in self._import_dict:
            return list(self._import_dict[load_from])
        else:
            # noinspection PyTypeChecker
            data_set_data: DataSetData = self.load_or_null(load_from, DataSetData)
            if data_set_data is None:
                raise Exception(f'Dataset with ObjectId={load_from} is not found.')
            if data_set_data.data_set != DataSourceData._empty_id:
                raise Exception(f'Dataset with ObjectId={load_from} is not stored in root dataset.')
            result = self.__build_data_set_lookup_list(data_set_data)
            self._import_dict[load_from] = result
            return list(result)

    @abstractmethod
    def _get_saved_by(self) -> Optional[ObjectId]:
        """Records where id_ is greater than the returned value will be
        ignored by the data source.
        """
        pass

    def __build_data_set_lookup_list(self, data_set_data: DataSetData) -> Optional[Set[ObjectId]]:
        if data_set_data is None:
            return

        if not ObjectId.is_valid(data_set_data.id_):
            raise Exception('Required ObjectId value is not set.')
        if data_set_data.key == '':
            raise Exception('Required string value is not set.')

        saved_by = self._get_saved_by()
        if saved_by is not None and data_set_data.id_ > saved_by:
            return

        result = set()
        result.add(data_set_data.id_)
        if data_set_data.imports is not None:
            for data_set_id in data_set_data.imports:
                if data_set_data.id_ == data_set_id:
                    raise Exception(f'Dataset {data_set_data.key} with ObjectId={data_set_data.id_} '
                                    f'includes itself in the list of its imports.')
                if data_set_id not in result:
                    result.add(data_set_id)
                    cached_import_list = self.get_data_set_lookup_list(data_set_id)
                    for import_id in cached_import_list:
                        result.add(import_id)
                else:
                    result.add(data_set_id)
        return result

    def get_data_set(self, data_set_id: str, load_from: ObjectId) -> ObjectId:
        result = self.get_data_set_or_none(data_set_id, load_from)
        if result is None:
            raise Exception(f'Dataset {data_set_id} is not found in data store {self.data_source_name}.')
        return result

    def get_common(self) -> ObjectId:
        return self.get_data_set(self.common_id, DataSourceData._empty_id)

    def create_common(self) -> ObjectId:
        result = DataSetData()
        result.data_set_name = self.common_id
        self.save_data_set(result, DataSourceData._empty_id)
        return result.id_

    def create_data_set(self, data_set_id: str, save_to: ObjectId, import_data_sets: List[ObjectId] = None) -> ObjectId:
        result = DataSetData()
        result.data_set_name = data_set_id

        if import_data_sets is not None:
            result.imports = [x for x in import_data_sets]

        self.save_data_set(result, save_to)
        return result.id_
